src/components/filters/TopCompoundFilter.jsx

```diff
@@ -0,0 +1,104 @@
+import React, { PropTypes } from 'react/addons';
+
+import CompoundFilter from './CompoundFilter';
+import ButtonSelect from '../inputs/ButtonSelect';
+
+import InputMixin from '../../mixins/InputMixin';
+
+const OPERATOR_OPTIONS = [
+  {
+    id: 'and',
+    label: 'AND',
+  },
+  {
+    id: 'or',
+    label: 'OR',
+  },
+];
+
+const defaultOperator = 'and';
+
+const TopCompoundFilter = React.createClass({
+
+  displayName: 'TopCompoundFilter',
+
+  mixins: [
+    InputMixin(PropTypes.shape({
+      // Id of the selected operator in `OPERATOR_OPTIONS`
+      operatorId: PropTypes.oneOf(OPERATOR_OPTIONS.map((operator) => operator.id)),
+      // List of the values of the `CompoundFilter` children
+      compoundFilters: PropTypes.arrayOf(CompoundFilter.PropTypes.value),
+    })),
+  ],
+
+  propTypes: {
+    filterInput: PropTypes.func,
+  },
+
+  _handleCompoundFilterChange(idx, value) {
+    this.requestChange({
+      compoundFilters: {
+        [idx]: { $set: value },
+      },
+    });
+  },
+
+  _handleAddCompoundFilter() {
+    this.requestChange({
+      compoundFilters: {
+        $push: [
+          {
+            operatorId: defaultOperator,
+            filters: [],
+          },
+        ],
+      },
+    });
+  },
+
+  _handleRemoveCompoundFilter(idx) {
+    this.requestChange({
+      compoundFilters: {
+        $splice: [[idx, 1]],
+      },
+    });
+  },
+
+  render() {
+    let { filterInput } = this.props;
+    let { compoundFilters } = this.getValue();
+
+    return (
+      <div className="TopCompoundFilter">
+        <div className="TopCompoundFilter-filters">
+          {compoundFilters.map((compoundFilter, idx) => (
+            <div key={idx} className="TopCompoundFilter-filter">
+              {/* The first compound filter shouldn't display the operator select, and isn't removable */}
+              {idx > 0 &&
+                <ButtonSelect
+                  className="TopCompoundFilter-operatorSelect"
+                  options={OPERATOR_OPTIONS}
+                  valueLink={this.linkValue('operatorId')}
+                  />
+              }
+              <CompoundFilter
+                filterInput={filterInput}
+                onRemove={compoundFilters.length > 1 ? this._handleRemoveCompoundFilter.bind(null, idx) : null}
```

Comment:
This is what I tried first, the issue with that is that `onRemove`'s prop type is function, and when `compoundFilters.length > 1` is false, `onRemove` will be a boolean, so the proptype validation will fail!